man/dbcat.1
.TH DBCAT 1 "12 September 1995"
.FB
.SH NAME
dbcat \- concatenate \fIFirstBase\fP databases
.SH SYNOPSIS
.B dbcat
[\fB-b\fP] [\fB-y\fP]
[\fB-o\fP \fIoutput\fP] \fIdatabase\fP [\fIdatabase\fP ...]
.SH DESCRIPTION
.I Dbcat
takes an arbitrary number of \fIFirstBase\fP databases and concatenates
them end to
end to create a single \fIFirstBase\fP database. All record totals are modified
accordingly.
.PP
If only one database is specified, \fIdbcat\fP(1)
will ask for another database name.
If an output file specifying the output database is not provided using
the \fB-o\fP flag, \fIdbcat\fP(1) will ask for an output file name.
The default name (achieved by hitting <RETURN> at the output file name prompt)
is \fIconcat\fP.
.PP
The \fB-b\fP flag will run the \fIFirstBase\fP tool in batch mode,
suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
The first database dictionary \fIdbcat\fP(1)
encounters
is assumed to be the shape of all subsequent databases.
Beyond checking that a file is actually a \fIFirstBase\fP database, 
no other error checking is done to determine whether the current
database matches the size and shape of the first one.
If a file is not a \fIFirstBase\fP database, an error message is given, and 
processing continues.
.PP
Note that \fIdbcat\fP(1) does not follow indexes, but concatenates
all undeleted records of each database to the output database.
.PP
\fIDbcat\fP(1)
allows an interesting tool oriented approach to data entry and data base
management systems.
Users can create personal databases that all have the same shape or
structure by using the same database dictionary. Then
each can work on their own database independent of others.
When all are done, the databases can be concatenated together to form
the union of all the databases.
.PP
Note: \fIdbclean\fP(1)
is automatically run at the completion of this program to
create a map file for the new database.
.SH FILES
.PD 0
.TP 10
*.cdb
\fIFirstBase\fP database. 
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary.
.TP 10
concat
default output file name for \fIdbcat\\fIfP\fP(1).
.PD
.SH SEE ALSO
dbcgen(1), dbedit(1), dbjoin(1), firstbase(5)
.PP
\fIFirstBase User's Guide and Reference Manual\fP
.br
.SH DIAGNOSTICS
If fed conflicting databases, \fIdbcat\fP(1) will die a terrible death.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbcat.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbcat.c,v 9.0 2001/01/09 02:55:53 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbcat_sid[] = "@(#) $Id: dbcat.c,v 9.0 2001/01/09 02:55:53 john Exp $";
#endif

/*  
 * Dbcat - concat databases together (end to end).
 *   read each record of each file and (if not deleted) output
 *   to concat output file. re-adjust the record count.
 */

#include <fb.h>
#include <fb_vars.h>

#define CONCAT_BASE	"concat"	/* dbcat dbase name */

extern char *cdb_DBCLEAN;
extern char *cdb_pgm;
extern short int cdb_secure;

static char ddict[FB_MAXNAME]    = {NULL}; /* for ddict file name */
static char msg[FB_MAXLINE], fname[FB_MAXNAME];

extern short int cdb_opendb_level;	/* in opendb - controls initlock */
extern int cdb_blockread;		/* block chars read - blockeach(3) */

int fd;
long wcount;

#if !FB_PROTOTYPES
static int concat();
static long scat();
static int cat_one();
static int finit();
static void dcopy();
static void usage();
#else /* FB_PROTOTYPES */
static int concat(int, char **);
static long scat(char *);
static int cat_one(fb_database *);
static int finit(char *, int, char **);
static void dcopy(char *);
static void usage(void);
extern int main(int, char **);
#endif /* FB_PROTOTYPES */


   main(argc, argv)
      int argc;
      char *argv[];
   
      {

         (void) Dbcat_sid;

         fb_getargs(argc, argv, FB_NODB);	/* just to set system */
         if (argc < 2)
	    usage();
         fb_scrhdr((fb_database *) NULL, "Get Names");
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
         if (concat(argc, argv) == FB_ERROR)
            fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	 if (!cdb_batchmode){
	    fb_move(4,1), fb_clrtobot();
	    fb_scrstat("Cleaning");
	    sprintf(msg, "Cleaning %s", fname);
	    fb_fmessage(msg);
	    }
	 sprintf(msg, "%s -b %s", cdb_DBCLEAN, fname);	/* borrow msg */
	 fb_system(msg, FB_WITHROOT);
	 fb_ender();
      }

/* 
 *  concat - concat all the fb_database together into fd.
 */
 
   static int concat(argc, argv)
      char *argv[];
      int argc;

      {
         long n;
	 int j, count;
	 char dname[FB_MAXNAME], xname[FB_MAXNAME];

	 for (count = 0, j = 1; j < argc; j++)
	    if (equal("-o", argv[j]))
	       j++;
	    else if (!(argv[j][0] == '-'))
	       count++;
	 if (count < 1)
	    usage();
	 else if (count == 1){
	    if (fb_getfilename(xname, "Second Database Name: ", "") == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
         fd = finit(fname, argc, argv);
	 fb_scrstat(fname);
	 fb_putseq(fd);			/* sequence a new fb_database */
	 fb_puthead(fd, 0L, 0L);
         if (cdb_secure)
            if (fb_putmode(fd, fb_getuid(), fb_getgid(), "666") == FB_ERROR)
               fb_xerror(FB_IO_ERROR, fname, NIL);
         fb_w_init(1, fd, -1);
	 if (!cdb_batchmode && !cdb_yesmode){
	    sprintf(msg, "Ok to Concat FirstBase Databases into %s (y/n)? ",
               fname);
	    if (fb_mustbe('y', msg, 23, 1) == FB_ERROR)
               fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
         n = 0L;
         if (!cdb_batchmode){
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
	 for (j = 1; j < argc; j++)
	    if (equal("-o", argv[j]))
	       j++;
	    else if (!(argv[j][0] == '-')){
	       fb_rootname(dname, argv[j]);
	       n += scat(dname);
	       }
	 if (count == 1){
	    fb_rootname(dname, xname);
	    n += scat(dname);
	    }
         fb_wflush(1);
	 if (fb_puthead(fd, n, 0L) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], NIL);
	 close(fd);
	 dcopy(fname);
	 return(FB_AOK);
      }

/* 
 *  scat - cat one file into fd. return number of recs catted.
 */
 
   static long scat(f)
      char *f;
      
      {
	 fb_database *db;
	 
	 db = fb_dballoc();
	 fb_mkstr(&(db->dbase), f);
	 fb_opendb(db, READ, FB_NOINDEX, 0);
	 strcpy(ddict, db->ddict);
	 fb_getco(db);
         fb_scrtime(db);
         wcount = 0;
         fb_blockeach(db, cat_one);
	 fb_closedb(db);
         cdb_opendb_level = 0;
         return(wcount);
      }

/* 
 *  cat_one - cat one record file into fd.
 */
 
   static int cat_one(dp)
      fb_database *dp;

      {
         if (fb_w_writen(0, dp->orec, cdb_blockread) == 0)
            fb_xerror(FB_WRITE_ERROR, cdb_pgm, NIL);
         wcount++;
         if (!cdb_batchmode){
            fb_gcounter(wcount);
            }
         return(FB_AOK);
      }

/* 
 * finit - merely check for existance of fname. get fname if not on
 *    argument line. ask permission to destroy if it exists.
 */
 
   static int finit(fname, argc, argv)
      char *fname, *argv[];
      int argc;
      
      {
         char str[FB_MAXLINE], outname[FB_MAXNAME];
	 int p;
	 
	 if ((p = fb_testargs(argc, argv, "-o")) == 0){
	    if (fb_getfilename(outname, "Output File Name: ", 
	          CONCAT_BASE) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 else
	    strcpy(outname, argv[p + 1]);
	 fb_rootname(fname, outname);
	 sprintf(outname, "%s.cdb", fname);		/* cdb extension */
	 if (access(outname, 0) != -1 && !cdb_batchmode && !cdb_yesmode){
	    sprintf(str,
	       "Permission to OVERWRITE %s (y = yes, <cr> = no)? ", fname);
	    if (fb_mustbe('y', str, 15, 10) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 close(creat(outname, 0666));
	 return(fb_mustopen(outname, READWRITE));
      }

/* 
 *  dcopy - copy dictionary from ddict to fname.ddict
 */
 
   static void dcopy(fname)
      char *fname;
      
      {
         FILE *in, *out;
	 char outname[FB_MAXNAME];
	 register int c;
	 
	 sprintf(outname, "%s.ddict", fname);
	 in = fb_mustfopen(ddict, "r");
	 out = fb_mustfopen(outname, "w");
	 while ((c = fgetc(in)) != EOF)
	    fputc((char) c, out);
	 fclose(in);
	 fclose(out);
      }

/* 
 *  usage message
 */
 
   static void usage()
      {
         fb_xerror(FB_MESSAGE, 
	    "usage: dbcat [-b] [-y] [-o output] dbase [dbase ...] ", NIL);
      }
